iroh-bytes/src/downloader.rs

```diff
@@ -454,540 +503,495 @@ struct Service<G: Getter, D: Dialer> {
+                    self.dialer.queue_dial(node);
+                }
+                NextStep::DialAfterIdleDisconnect(node, key) => {
+                    let expired = self.goodbye_nodes_queue.remove(&key);
+                    let expired_node = expired.into_inner();
+                    debug!(node=%expired_node.fmt_short(), "disconnect idle node to make room for next connection");
+                    let info = self.nodes.remove(&expired_node);
+                    debug_assert!(
+                        matches!(
+                            info,
+                            Some(ConnectionInfo {
+                                state: ConnectedState::Idle { .. },
+                                ..
+                            })
+                        ),
+                        "node picked from goodbye queue to be idle"
+                    );
+                    debug!(node=%node.fmt_short(), kind=%kind.fmt_short(), "dial node");
+                    self.dialer.queue_dial(node);
+                }
+                NextStep::OutOfProviders => {
+                    debug!(kind=%kind.fmt_short(), "abort download: out of providers");
+                    let _ = self.queue.pop_front();
+                    let info = self.requests.remove(&kind).expect("queued downloads exist");
+                    self.finalize_download(kind, info.intents, Err(DownloadError::NoProviders));
                 }
             }
+        }
+    }
+
+    fn next_step(&self, kind: &DownloadKind) -> NextStep {
+        if self
+            .concurrency_limits
+            .at_requests_capacity(self.active_requests.len())
+        {
+            return NextStep::Wait;
         };
 
-        // we tried to get a node to perform this request but didn't get one, so now this attempt
-        // is failed
-        if remaining_retries > 0 {
-            remaining_retries -= 1;
-            self.schedule_request(kind, remaining_retries, next_node, intents);
-        } else {
-            // check if this hash is needed in some form, otherwise remove it from providers
-            let hash = *kind.hash();
-            if !self.is_needed(hash) {
-                self.providers.remove(hash)
+        let mut candidates = self.providers.get_candidates(&kind.hash()).peekable();
+        if candidates.peek().is_none() {
+            return NextStep::OutOfProviders;
+        }
+
+        let mut available = vec![];
+        let mut currently_dialing = 0;
+        let mut has_exhausted = false;
+        let mut next_to_dial = None;
+        for node in candidates {
+            match self.node_state(node) {
+                NodeState::Connected(info) => {
+                    let active_requests = info.active_requests();
+                    if self
+                        .concurrency_limits
+                        .node_at_request_capacity(active_requests)
+                    {
+                        has_exhausted = true;
```

Comment:
Added this comment to the code:
```       
         // Track if we have at least one provider node which is currently at its request capacity.
        // If this is the case, we will never return [`NextStep::OutOfProviders`] but [`NextStep::Wait`]
        // instead, because we can still try that node once it has finished its work.
```
does this make it clear? if not please post back and I'll refine the docs

Reply:
this makes it clear